### ISSUE TITLE
(Google) Map Field

### ISSUE THREAD

**Issue by medienbaecker** [k-next/plugins#3]
For some projects a map field would be great to have.
I recently used https://github.com/hdodov/kirby-simplemap-field which works great in Kirby 2.

**Comment by rasteiner** [comment]
actually, I'd prefer this one: https://github.com/rasteiner/kn-map-section
development is currently on ice until section plugins API has finalized. 

I think the section is much cleaner approach, because it's only a panel visualization and you can choose what to store (center coords? zoom level? markers? markers with what fields?) and where.

**Comment by REHvision** [comment]
:+1: :+1: :+1: 

Actually, I would love to choose between map providers (Google Maps and OpenStreetMap/Leaflet for example).

What do you think, should there be only ONE maps (section) plugin that wraps different map solutions and config options or should they be individual plugins?

Plugin names like **Map**, **Maps**, **Map-Section** etc. sound like general mapping solutions.
I think the plugin name should reflect that it's Google Maps (if thats the only option and no other providers are planned). :smile:

**Comment by bvdputte** [comment]
There's also: 
https://github.com/lekkerduidelijk/kirby-geolocation-field

It's quite useful imho when you only want to save the lat/lng as field values. In the template/front end you can then decide what you want to do with it. E.g. I've built a "service station locator" with it: https://mullion-pfd.com/en/service-stations -> Each service-station is a page which uses that field to save the lat/lng which is then used in the front end to feed the locator Javascript.

**Comment by wottpal** [comment]
Why not just completely 💩 on Google and make the OpenStreetMap solution the default? I've just found [Leaflet](https://leafletjs.com) which sounds very promising and also already has a vue-component available: [Vue2Leaflet](https://www.npmjs.com/package/vue2-leaflet).

**Comment by bvdputte** [comment]
Imho it might be useful to have 

- 1 plugin which makes "selecting a location" in the panel the best as possible (lat/lng is "provider"-independent anyway)
- 1 plugin which builds on that (loosely coupled), which makes it easy to display it in the template/front end

**Comment by rasteiner** [comment]
I think that *my* will come bundled with google maps because here where I (and my clients) live it's the most complete mapping solution. The google api client is loaded dynamically so doesn't have to add to the bundle size.

What I could do is generalize the interface to it so that another plugin can inject it's map provider as a vue component. It would need to accept props like center, zoom and markers and emit events when those change.

**Comment by bnomei** [comment]
- while i like the technical solution @rasteiner has i and my clients would dearly miss styling the map. if interested i would like to join forces on this – otherwise i will port my existing V2 plugin: https://github.com/bnomei/kirby-styledmap
- with some careful planing (and use of snippets as templates) one could support leaflet and google api at the same time.

**Comment by rasteiner** [comment]
I think, like bvdputte said before, we need to distinguish between a map visualization in the panel (section or field) and maps we show on frontend. 
I, too, think these should be 2 different plugins (and both could work with or without the other). After all it's just coordinates saved in a textfile ;)

Map styling would be part of the frontend plugin.

**Comment by bnomei** [comment]
but i case of my plugin its not one coordiate but multiple on one map. each with title, popup hover text, link and custom icon support. thats also data to store somewhere.

**Comment by rasteiner** [comment]
yup. In case of kn-map-section, markers are saved in any field that accepts arrays (like the structure field). Every structure item has then any field you want. :)

If the styled map plugin is built around accepting a structure field (or any array) they might just work together. In my plugin there's the convention of naming markers with a field named "title" and their coordinates in a field called "coords".

**Comment by bnomei** [comment]
i am willing to help on this one so i am assigning myself. shall we flag it like this?

**Comment by rasteiner** [comment]
I'll refactor my plugin next week to allow other map providers. I'll open a ticket there: https://github.com/rasteiner/kn-map-section/issues/1

**Comment by fourpixels** [comment]
I like @bvdputte's idea to have something super basic and another one to reuse the base one and build upon with all the extra features.

Either way I haven't checked all repositories (turned out there are quite a few! 😋 But I think we can make some simple solution that works for everyone, as long as we don't want it to solve completely everything..
Keep up the good work guys!

**Comment by luxlogica** [comment]
The ideal solution that the developer needs is this:

- client enters the location address in Panel, and ideally sees immediately a preview of how the address will be interpreted and displayed by Google.
- client is able to select zoom level in Panel - and also preview the changes in the Panel immediately.
- developer can then use a helper function or field method to directly output the required embedded map on the frontend - e.g., `<?= $page->my_geofield()->map() ?>`

Bonus points if we could also have a config option to use **[SnazzyMaps](https://snazzymaps.com)** code to customise the map's appearance.

**Comment by rasteiner** [comment]
closed https://github.com/rasteiner/kn-map-section/issues/1

you'll find some instructions on how to inject your map component there. 
For people that used the section before, the blueprint has changed a bit (you now reference directly a fieldname in `storage`). 

I've also removed the `geocoded` field that was included (a field that transformed addresses into coordinates) because that used google, too. It seemed wrong to expect your map components to also provide geocoding functionality. 
I'll release it as a separate plugin.
